Improve the CLI and other code improvements

theRookieCoder opened the pull request.

theRookieCoder committed:
Cli (#1)

* Rewrite CLI

* CLI documentation, remove empty jar,
and other improvements

* Bump deps

theRookieCoder committed:
Improve CLI and its code ergonomics

theRookieCoder committed:
More improvements

theRookieCoder committed:
Switch back to glow due to issues on macOS

theRookieCoder committed:
changes

theRookieCoder committed:
Code improvements

theRookieCoder committed:
Update deps

CheaterCodes commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
-
-        #[arg(short, long)]
+        /// Don't create a profile
+        #[arg(short = 'p', long)]
```
While you're at it:
I think doing this with `--create-profile false` (or `-p false`), with `true` being the default value would be more intuitive.

theRookieCoder replied:
That's what I did initially, however that means that if you want to disable profile creation, you'd have to do `--create-profile false` rather than `--no-profile`. Basically I feel like the false is redundant given that true is the default anyways. Also CLI programs generally follow the latter pattern for the same reason.

UpcraftLP replied:
I agree with @theRookieCoder here, but perhaps change the short option to a capital `P` to denote that it is negated

CheaterCodes commented on src/installer.rs:
```diff
@@ -143,28 +153,26 @@ pub async fn install_client(client: Client, args: ClientInstallation) -> Result<
-        // This really shouldn't be fixed here in the installer, but we need a solution now.
-        let mut json: Value = serde_json::from_str(&response).unwrap();
+        // Quilt-meta specifies both hashed and intermediary,
+        // but providing both to quilt-loader causes it to silently fail remapping.
```
maybe specify that this applies to old versions only.

theRookieCoder replied:
It is at [line 171](https://github.com/QuiltMC/quilt-native-installer/pull/20/files#diff-4904922dab3fd7bd66e213f49476ad3aa39a8a7c0cb24f68a68343909bbdaab4R171)

theRookieCoder committed:
Update to iced `0.9`
Rearrange iced imports
Remove unnecessary alignments
Disable output dir text input fields when installing
Other small code improvements

theRookieCoder committed:
Bump deps

theRookieCoder committed:
Update dependencies

theRookieCoder committed:
Update dependencies

UpcraftLP commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
-        install_dir: PathBuf
-    }
+        /// Do not generate launch scripts
+        #[arg(short = 's', long)]
```
perhaps change the short option to a capital `S` to denote that it is negated

UpcraftLP commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
+        #[arg(short = 's', long)]
+        no_script: bool,
+        /// Do not download the server jar
+        #[arg(short = 'j', long)]
```
perhaps change the short option to a capital `J` to denote that it is negated

UpcraftLP commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
-    }
+        /// Do not generate launch scripts
+        #[arg(short = 's', long)]
+        no_script: bool,
```
`--no-script` should probably be changed to `--no-launch-scripts`

UpcraftLP commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
+            MCVersionCLI::Custom(input) => minecraft_versions
+                .into_iter()
+                .find(|v| v.version == input)
+                .context(format!("Could not find Minecraft version {}", input))?,
```
this should use [`with_context()`](<https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.with_context>) instead:

```rs
.with_context(|| format!("Could not find Minecraft version {}", input))?
```

UpcraftLP commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
+            LoaderVersionCLI::Custom(input) => loader_versions
+                .into_iter()
+                .find(|v| v.to_string() == input)
+                .context(format!("Could not find Quilt Loader version {}", input))?,
```
this should use [`with_context()`](<https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.with_context>) instead

UpcraftLP commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
+            LoaderVersionCLI::Stable => loader_versions
+                .into_iter()
+                .find(|v| v.version.pre.is_empty())
+                .unwrap(),
```
there should be a proper error message here

UpcraftLP commented on src/cli.rs:
```diff
@@ -1,109 +1,170 @@
+            LoaderVersionCLI::Beta => loader_versions
+                .into_iter()
+                .find(|v| !v.version.pre.is_empty())
+                .unwrap(),
```
there should be a proper error message here

UpcraftLP reviewed (changes requested):
Looks nice! I've left a few minor nitpicks tho.

theRookieCoder committed:
Implemented Up's review suggestions

Co-Authored-By: Up <upcraftedlp@gmail.com>

UpcraftLP commented on src/cli.rs:
```diff
@@ -150,21 +150,21 @@ async fn get_versions(
                 .into_iter()
                 .find(|v| v.version.pre.is_empty())
-                .unwrap(),
+                .with_context(|| "Could not find a stable Quilt Loader version")?,
```
this on the other hand should only be `context()` because there is no point in lazily evaluating a constant.

theRookieCoder replied:
Oh 🤦 I was mindlessly copy-pasting and missed that. Usually with this type of mistake clippy catches it, but it didn't this time I think because it's a third-party library.

UpcraftLP commented on src/cli.rs:
```diff
@@ -150,21 +150,21 @@ async fn get_versions(
                 .into_iter()
                 .find(|v| !v.version.pre.is_empty())
-                .unwrap(),
+                .with_context(|| "Could not find a beta Quilt Loader version")?,
```
same as above.

theRookieCoder committed:
Remove lazy loading for static string

UpcraftLP reviewed (approved):
LGTM 👍🏻
